### ISSUE TITLE
Insufficient supporting arguments are given for why "virtual void method() override" should be avoided

### ISSUE THREAD

**Issue by sean-mcmanus** [isocpp/CppCoreGuidelines#1271]
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final

Insufficient supporting arguments are given for why "virtual void method() override" should be avoided. The "virtual" is of course not required, but our team believes it increases readability to always have virtual on the left for virtual methods (the "override" on the right might be scrolled off the screen or on a different line). If "override" could be moved over to the left like in C# that would be preferable (I assume it would break compatibility). Is someone able to add some convincing examples or additional reasons why the extra virtual should be avoided?

**Comment by scraimer** [comment]
Anecdote: I've watched an inexperienced programmer remove the "virtual" from a function declaration, and was surprised it didn't make any difference.

That sort of ambiguity is partly responsible for reputation of C++ as a difficult-to-learn language.

Imagine a novice developers browsing a code base where this rule is enforced. Every member function has only one of "virtual", "override" or "final". Then each function belongs to exactly one of the following categories:

* Doesn't participate in the inhertence (none of the keywords are present)
* Is the base-most declaration of the member function ("virtual")
* Is an inherting and overriding member function ("override", which implies "virtual")
* Is the child-most declaration of the member function ("final", which may imply "override" if there was an ancestor class which could be overridden)

In code that doesn't enforced it, you can run into confusing situations:

* If a function should have been marked override but wasn't, it's hard to know. In the past, such functions were marked with "virtual", or left with nothing. If marked with nothing, and the base class had that member function removed or changed (arguments or name changes), often there would be no compile error, but the wrong function might be called in the code. If marked with "virtual", it would not be clear if that function was overriding a member of a base class or establishing a new chain of inhertence.

* If a function was marked "virtual" when it was supposed to be marked "override", that's also confusing.

In short, enforcement of the rule is better for experssiveness.

**Comment by sean-mcmanus** [comment]
Yeah, our code follows the enforceable rules.
1. Nothing
2. virtual
3. virtual override
4. virtual final

We just add the duplicate virtual. I still don't understand how the duplicate virtual can lead to any more confusion or errors compared to omitting it. I understand the problems solved by override/final, but I don't understand how using the extra virtual increases the chance of error or confusion -- someone could still use virtual when they meant override, the same as if they use virtual when they meant virtual override. A specific, compelling example that could be used to convince other people would help.

@bobrow @grdowns Do you guys have any thoughts on this?

**Comment by bobbrow** [comment]
If `override` came before the declarator, I would be fine with removing the duplicate `virtual`.  The reason I prefer to keep both is because my brain has been trained to look for the virtual keyword before the declarator instead of after.  C++ chose to put the `override` keyword at the end of the declaration for some reason and if you remove the `virtual` keyword for overrides, that breaks having a consistent place to look for virtual functions when reading someone else's code.

**Comment by hsutter** [comment]
Personally, my view is that writing exactly one of them is a simple rule to teach and desirable for code clarity for the reasons @scraimer mentioned.

I do understand that some people may say "but why can't I write virtual if I also write override?" FWIW, in my experience (which may be a limited sample of course) I much more commonly hear "but why should I have to write virtual if I already wrote override?"

The only annoyance is that C++ doesn't allow them all in the same lexical position (oh well). But I personally don't think that's enough to justify the duplication and verbosity.

That said, I would be willing to consider making the rule "write _at least one of_ virtual/override/final." Although I find "exactly one" easier to teach and learn, the 90% value in this area is to avoid implicit overrides and either formulation of the rule would accomplish that.

**Comment by cubbimew** [comment]
> "write at least one of virtual/override/final." 

Formulated this way, it brings back to the kind of bugs override prevents:

    struct Base                { virtual int foo(int n)           { return n; } };
    struct Derived : Base { virtual int foo(int n) const { return n*2; } };

**Comment by hsutter** [comment]
@cubbimew Yup, you're right. The alternative rule I mentioned only covers the "implicit virtual (or is it?)" pitfall, it does not cover the "implicit override (or is it?)" pitfall.

So I don't readily see an alternative to "write exactly one of virtual/override/final" that is both (a) simple, and (b) eliminates the "implicit virtual/override (or is it?)" problems.

Now that you've reminded me, one thing I always liked a lot about "write exactly one" is that you just can't make any of those mistakes, including the one I just made again of writing virtual and forgetting to write override. "After all, I wrote something!" but it wasn't the right thing.

**Comment by bobbrow** [comment]
Thanks for the update. I do appreciate the clarifications in that commit.

I can certainly understand the teachability argument. And the rationale for requiring `override`/`final` is sound - we are enforcing this. The deal breaker (for me) in changing our codebase to use *just one* continues to be the opposing lexical position of the keywords.  The lexical position of `virtual` allows me - the code reader/reviewer - to more easily find virtual methods and then I can scan to the right to check for an `override` keyword to complete my context and enforce a variant of this guideline.  For me, the guideline is more about enforcement of `override`/`final`.

**Comment by sean-mcmanus** [comment]
I am still not understand why "at least one" is not sufficient.
@cubbimew Your example only uses one virtual without a 2nd modifier, so I don't see how leaving open the potential for an override or final would lead to any problem.
@hsutter The mistake of writing virtual without the override could still occur if the virtual is omitted when override is used.

It seems as if the argument is that adding the duplicate virtual would increase the likelihood of a mistakenly omitted override? -- I don't find that argument convincing. One could just as easily forget the virtual and override and the results would be exactly the same (i.e. the base class method getting called).

What I'm looking for is an example in which a tool could be used to identify the incorrect usage only when exactly one modifier is used -- I don't see how would be possible.

**Comment by hsutter** [comment]
It's not just about the tool, it's about human readability (e.g., code review) and reducing the opportunities to make mistakes. YMMV, but I like this rule and find it very helpful.

**Comment by hsutter** [comment]
@bobbrow Would your opinion change if C++ allowed `virtual` at the end (where `override` and `final` are allowed) and/or allowed `override` and `final` at the front (where `virtual` is allowed)? That is, is the fact that they are in different positions your primary concern?

Disclaimer: This is an exploratory question to better understand your concern, I'm not saying I would propose that change to C++.

**Comment by bobbrow** [comment]
> Would your opinion change if C++ allowed virtual at the end (where override and final are allowed) and/or allowed override and final at the front (where virtual is allowed)?

Absolutely.  If I could put `override` at the beginning of the declaration I feel there would be no need for `virtual` to coexist on that declaration.

> Disclaimer: This is an exploratory question to better understand your concern, I'm not saying I would propose that change to C++.

Fair enough. 😄

**Comment by hsutter** [comment]
@bobbrow Thanks. Just to be clear, re-asking the other half of my question: Would your opinion change if you could write virtual at the end (whether or not we ever allowed writing override/final at the beginning)? I'm probing to see if the position matters to you too, or just the togetherness.

**Comment by bobbrow** [comment]
I have a strong preference for writing `override`/`final` at the beginning, but I don't think it would be a deal breaker if `virtual` were allowed at the end instead.  I think the togetherness is more important than the positioning.

The reason for my preference is that I do a fair bit of C# programming as well so I prefer that the two languages have similar placement of similar keywords so I don't have to do as many mental gymnastics when switching between languages.

**Comment by hsutter** [comment]
@bobbrow OK, thanks for your input. (If people agree/disagree feel free to up/down-thumb @bobbrow's last two replies.)

**Comment by hsutter** [comment]
(Whatever you do, please don't make a macro for

~~~
VIRTUAL(void f());
OVERRIDE(void f());
FINAL(void f());
~~~

which doesn't cover all language quirks and is just so evil in principle that I would never mention it even in passing lest it risk planting the idea... oh, wait...)

**Comment by jwakely** [comment]
> I am still not understand why "at least one" is not sufficient.

@sean-mcmanus because as cubbimew said, "at least one" means if you say `virtual` then you don't need `override` (because `virtual` is one, and one is at least one, so you don't need any more). If you don't put `override` then you don't get the checking that `override` gives, and you are back to fragile classes that accidentally add another overload instead of overriding a virtual function from the base class.

**Comment by scraimer** [comment]
Aside: It seems I missed a case. You'll have to specify both (`override final`) if you want to mark the function as `final` and also gain the protection of the `override`.

At least, that's what I'm getting from my reading of the standard. Please correct me if I'm wrong.

**Comment by scraimer** [comment]
@sean-mcmanus

> What I'm looking for is an example in which a tool could be used to identify the incorrect usage only when exactly one modifier is used -- I don't see how that would be the case.

I have a case where the additional `virtual` was harmful, and if we had used only a single specifier, we would have gotten a compiler error instead of not being notified about it. I'm not sure it applies to you. It happened to me during the transition to modern C++. In particular, this was before all the functions were correctly marked `override`.

Start with this set of classes, where we always specify `virtual`:

    class A {
       virtual int foo() { return 1; };
    };

    class B : public A {
       virtual int foo() { return 2; }
    };

    class C : public B {
       virtual int foo() override { return 3; }
    };

If you have an `A* a`, then the expression `a->foo()` would call either `A::foo` or `B::foo` or `C::foo`, depending on whether `a` was pointing to `A` or `B` or `C`.

But then the signature of `A::foo` changed. It now would accept an argument. Actually, it was an optional argument. (The optional argument just made code review a little harder.)

    class A {
       virtual int foo(int val = 0) { return 4; };
    };

The compiler gave no error or warning, because:
* `B::foo` was not marked `override`
* `B::foo` was marked `virtual`, so `C::foo` was still able to be marked `override`

The result:
At run time, `a->foo()` only ever called `A::foo`. Regardless of whether `a` pointed to `A` or `B` or `C`.

If the additional `virtual` in `B::foo` would not have been there, we would have gotten a compiler error.

**Comment by brenoguim** [comment]
@scraimer , but according to the rule where we add `virtual` to improve readability mentioned in the beginning of the thread, you have to make the function either `virtual` or `virtual + override` or  `virtual + final`.
In the case you mention, you are not getting a bug because of following this rules, but because you failed to follow it by not adding `override` on `B::foo`.

**Comment by jwakely** [comment]
@scraimer 

> Aside: It seems I missed a case. You'll have to specify both (override final) if you want to mark the function as final and also gain the protection of the override.

It is ill-formed to use the virt-specifier `final` on a non-virtual function, so if you use it _without_ `virtual` then you get checking (it will be ill-formed if it fails to override a virtual function from the base class).

If you put `virtual` _and_ `final` then you get no checking, because it's a virtual function.

**Comment by brenoguim** [comment]
> If you put virtual and final then you get no checking, because it's a virtual function.

And that is the reason for not making things `virtual + final`, right?

If you make a method `virtual + final`, it will compile even if it does not override any other method, while if you only mark it only `final`, you will get the compiler error.
This is a strong case for using only one of the words I guess.

**Comment by jwakely** [comment]
Yup.

So those who want to add `virtual` redundantly (the so-called "at least one" rule) would really need the rule to be "use the maximum possible number of keywords"! Because as I explained previously, `virtual` on its own doesn't check, so you want `virtual override` if it's an override. And as just discussed, `virtual final` also doesn't check, so you want `virtual final override` on a final overrider.

So the discussion is whether to prefer the current guideline:

- `virtual` means exactly and only "this is a new virtual function."
- `override` means exactly and only "this is a non-final overrider."
- `final` means exactly and only "this is a final overrider."

or the "use `virtual` on all virtual functions" which would be:

- `virtual` means exactly and only "this is a new virtual function."
- `virtual override` means exactly and only "this is a non-final overrider."
- `virtual final override` means exactly and only "this is a final overrider."

There's also another case not covered by either, which is a virtual function which is final but not overriding anything. I guess both rules would use `virtual final` for that (because there's no other way to say it).

**Comment by hsutter** [comment]
Editors call: We agree with @jwakely's comment. It seems that this thread has bottomed out and we like the way the guideline is currently stated. Thanks for helping us to write it more clearly though!
